src/message_bus.zig

```diff
@@ -1,55 +1,228 @@
 const std = @import("std");
 const assert = std.debug.assert;
-const Allocator = std.mem.Allocator;
+const mem = std.mem;
+const os = std.os;
 
 const vr = @import("vr.zig");
 const ConfigurationAddress = vr.ConfigurationAddress;
 const Header = vr.Header;
 const Replica = vr.Replica;
 const FIFO = @import("fifo.zig").FIFO;
+const IO = @import("io_callbacks.zig").IO;
 
 const log = std.log.scoped(.message_bus);
 
+const tcp_backlog = 64;
+const num_connections = 32;
+
 pub const Message = struct {
     header: *Header,
     buffer: []u8 align(vr.sector_size),
-    references: usize = 0,
+    references: usize = 1,
     next: ?*Message = null,
 };
 
+const Envelope = struct {
+    message: *Message,
+    next: ?*Envelope = null,
+};
+
+// TODO: use a hasmap to make client lookups faster
 pub const MessageBus = struct {
-    allocator: *Allocator,
+    allocator: *mem.Allocator,
     allocated: usize = 0,
-    configuration: []ConfigurationAddress,
+    io: *IO,
+
+    configuration: []std.net.Address,
+
+    /// The replica which is running the server
+    server: *Replica,
+    server_fd: os.socket_t,
+    /// Used to store messages sent by the server to itself for delivery in flush().
+    server_send_queue: FIFO(Envelope) = .{},
 
-    /// Messages that are ready to send:
-    ready: FIFO(Envelope) = .{},
+    accept_completion: IO.Completion = undefined,
+    accept_in_progress: bool = true,
 
-    const Address = union(enum) {
-        replica: *Replica,
-    };
+    /// This slice is allocated with a fixed size in the init function and never reallocated.
+    connections: []Connection,
+    /// Number of connections currently in use (i.e. connection.peer != .none).
+    connections_in_use: usize = 0,
```

Comment:
I think it's slightly inaccurate as connections may be "in use" without being connected while they are shutting down. However, this is equally true when talking about replicas...

Reply:
Yes, I was thinking the same. We could do `replica_connections_in_use` to bring it inline the other way round.

But even better would be to get rid of the `replicas_connected` count and replace this with a `replicas_fully_connected` boolean that we use as a dirty bit, because that's all we need in `tick()`. That avoids any misunderstanding like I had around off-by-one, it's easier to not get the dirty bit wrong compared to a count, and it suits the purpose we need it for.

Let's then do `connections_used`.

Reply:
As discussed in the meeting this morning, I've removed this check and the `replicas_connected` count entirely for now in the interest of simplicity. This can be revisited when/if we need marginally more performance out of `tick()`.